### ISSUE TITLE
Change `kani::any::<bool>()` to make the non-det bit pattern more intuitive

### ISSUE THREAD

**Issue by sanjit-bhat** [model-checking/kani#1490]
## TLDR

This issue proposes changing our implementation of `kani::any::<bool>()` from:

https://github.com/model-checking/kani/blob/de71c097f266053d6b10c1c4d762b47766a210d9/library/kani/src/arbitrary.rs#L48-L54

to something like

```rust
impl Arbitrary for bool {
    #[inline(always)]
    fn any() -> Self {
        let byte = u8::any();
        crate::assume(byte < 2);
        byte == 1
    }
}
```

## Reasoning

Our current implementation of `kani::any::<bool>()` has the correct API semantics of a `bool`. I.e., that a `bool` can be either `true/false`. However, it doesn't map to the [underlying bit pattern of Rust bools](https://doc.rust-lang.org/reference/types/boolean.html):

> An object with the boolean type has a [size and alignment](https://doc.rust-lang.org/reference/type-layout.html#size-and-alignment) of 1 each. The value false has the bit pattern 0x00 and the value true has the bit pattern 0x01. It is [undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) for an object with the boolean type to have any other bit pattern.

The Rust spec has the following differences:
1. A `true` value is represented by `0x01`, not `0x00`.
2. A `false` value is represented by only `0x00`, not `[0x01, 0xFF]`.

This difference has an impact when playing back deterministic test cases. In our tests for playing back `bool`'s, we can never set an expected deterministic bit pattern for `false`. There are multiple possible bit patterns, and we don't know which one the solver will choose.

## Callouts

The proposed change would add an extra `kani::assume()` for every `bool` generated. This might slow down verification time. We should performance benchmark this change before implementing it.

**Comment by tedinski** [comment]
I don't quite get the "reasoning" section. The function returns only true or false (the result of the `==`). The bits of the `u8` don't really matter except that they're consistently interpreted.

Test replay should see whatever bits for the `u8` and just continue with the `byte == 1` test, returning true or false.

**Comment by sanjit-bhat** [comment]
@tedinski you're right that the bits are consistently interpreted.

For the test replay issue, I'm referring to the test here:

https://github.com/model-checking/kani/pull/1488/files#diff-8fa480379b673486e8cfa50d3abc61f0c09dfc6fe97afd2a4b4fb8531a036807R9-R10

with expected output here:

https://github.com/model-checking/kani/pull/1488/files#diff-0f14a3452f9e948e47159f2f28eddffe6f7a951d8c87bea823ba980045a55906R7-R10

This test asserts that the specific bit patterns are correct, and it doesn't currently work for `false` values.

**Comment by zhassan-aws** [comment]
This sounds to me like a bug/limitation in the executable trace implementation? The underlying `u8` value is not supposed to be exposed outside `kani::any::<bool>()`.

**Comment by sanjit-bhat** [comment]
It's a limitation in the test cases for the executable trace feature. This limitation might be removed in the future once I add  additional test functionality to run the generated deterministic tests and check that the interpreted values are correct.

However, is there any reason why we shouldn't do this? Have we tested the performance of the alternate implementation? To me, it seems like we should be doing it the way of the alternate implementation (it's closer to how Rust does `bools`), unless there's a significant performance impact.

**Comment by celinval** [comment]
Do you mind if we edit the tittle? I don't think this is not a bug and kani::any::<bool>() does respect the actual Rust bool bit pattern.

That said, if there is no performance impact, I don't mind changing this to match 0 / 1. I do agree that it might make it simple for users to read the trace (with --visualize) as well as the executable trace implementation.

**Comment by celinval** [comment]
Here is a test that shows that `kani::any::<bool>` does respect the Rust bool bit pattern:

```rust
#[kani::proof]
fn check_bool_bits() {
    let b : bool = kani::any();
    let bits = b as u8;
    assert!(bits == 0x0 || bits == 0x1);
}
```

**Comment by sanjit-bhat** [comment]
Hm, that verifies, even if I replace `as` with `transmute`.

I'm confused why that's the case. I thought the underlying bit pattern (the `let byte = u8::any()`) can be any arbitrary `u8`. Why is it that these `asserts` work out?

**Comment by zhassan-aws** [comment]
> I thought the underlying bit pattern (the `let byte = u8::any()`) can be any arbitrary `u8`.

It definitely can be.

> Why is it that these `asserts` work out?

Because `kani::any::<bool>()` does not return `byte`; it returns `byte == 0` which can only be `0x00` or `0x01`.

**Comment by sanjit-bhat** [comment]
Ah, that makes sense now. Thanks everyone for helping to clarify that!

@tedinski @celinval @zhassan-aws I updated the title to reflect the fact that this issue is not a problem with the `kani::any::<bool>()` bit pattern. Rather, it's a usability improvement. If users inspect the output of `--visualize` or `--gen-exe-trace`, this change will make the non-det byte used to generate `bool`s look more like the byte underlying Rust `bool`s.

**Comment by sanjit-bhat** [comment]
## Performance eval of switching to new `any::<bool>()`

Configurations:
1. rectangle example: stretched_rectangle_can_hold_original_fixed
2. firecracker: requirement_2642
3. vecdeque-cve: minimal_example_with_cve_fixed

For each configuration, we show numbers for a) verification time, b) elapsed (wall clock) time, and c) max resident set size. b) and c) are from `gtime -v`, where `gtime` is the GNU `time` program.

### Old bool
1. rect example: 0.08s, 1.19s, 106mb
2. firecracker: 3.06s, 4.77s, 127mb
3. vecdeque-cve: 0.37s, 3.51s, 153mb

### New bool
1. rect example: 0.09s, 1.25s, 110mb
2. firecracker: 3.13s, 4.88s, 132mb
3. vecdeque-cve: 0.39s, 3.62s, 146mb
